bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVeneerSettingsFactory.java

```diff
@@ -0,0 +1,327 @@
+        .setMaxAttempts(retryOptions.getMaxScanTimeoutRetries());
+
+    // configurations for RPC timeouts
+    Duration readPartialRowTimeout = ofMillis(retryOptions.getReadPartialRowTimeoutMillis());
+    retryBuilder
+        .setInitialRpcTimeout(readPartialRowTimeout)
+        .setMaxRpcTimeout(readPartialRowTimeout)
+        .setTotalTimeout(ofMillis(options.getCallOptionsConfig().getLongRpcTimeoutMs()));
+
+    builder.readRowsSettings()
+        .setRetrySettings(retryBuilder.build())
+        .setRetryableCodes(buildRetryCodes(options.getRetryOptions()));
+  }
+
+  /** Creates {@link RetrySettings} for non-streaming idempotent method. */
+  private static RetrySettings buildIdempotentRetrySettings(RetrySettings retrySettings,
+      BigtableOptions options) {
+    RetryOptions retryOptions = options.getRetryOptions();
+    RetrySettings.Builder retryBuilder = retrySettings.toBuilder();
+
+    retryBuilder
+        .setInitialRetryDelay(ofMillis(retryOptions.getInitialBackoffMillis()))
+        .setRetryDelayMultiplier(retryOptions.getBackoffMultiplier())
+        .setMaxRetryDelay(ofMillis(retryOptions.getMaxElapsedBackoffMillis()))
+        .setMaxAttempts(retryOptions.getMaxScanTimeoutRetries());
+
+    // configurations for RPC timeouts
+    Duration shortRpcTimeout = ofMillis(options.getCallOptionsConfig().getShortRpcTimeoutMs());
+    retryBuilder
+        .setInitialRpcTimeout(shortRpcTimeout)
+        .setMaxRpcTimeout(shortRpcTimeout)
+        .setTotalTimeout(ofMillis(options.getCallOptionsConfig().getLongRpcTimeoutMs()));
+
+    if (retryOptions.allowRetriesWithoutTimestamp()) {
+      LOG.warn("Retries without Timestamp does not support yet.");
+    }
+
+    return retryBuilder.build();
+  }
+
+  /** Creates {@link Set} of {@link StatusCode.Code} from {@link Status.Code} */
+  private static Set<StatusCode.Code> buildRetryCodes(RetryOptions retryOptions) {
+    ImmutableSet.Builder<StatusCode.Code> statusCodeBuilder = ImmutableSet.builder();
+    for (Status.Code retryCode : retryOptions.getRetryableStatusCodes()) {
+      statusCodeBuilder.add(GrpcStatusCode.of(retryCode).getCode());
+    }
+
+    return statusCodeBuilder.build();
+  }
+
+  /** Creates {@link TransportChannelProvider} based on Channel Negotiation type. */
+  private static TransportChannelProvider createChannelProvider(String hostname,
+      BigtableOptions options) {
+    final String endpoint = hostname + ":" + options.getPort();
+    String userAgent = VENEER_ADAPTER + options.getUserAgent();
+
+    //InstantiatingGrpcChannelProvider has special handling for User-Agent header. It set the
+    //ChannelBuilder with provided value.
+    HeaderProvider headers = FixedHeaderProvider.create(USER_AGENT_KEY.name(), userAgent);
+
+    InstantiatingGrpcChannelProvider.Builder builder =
+        InstantiatingGrpcChannelProvider.newBuilder()
+            .setChannelsPerCpu(2)
+            .setHeaderProvider(headers)
```

Comment:
Shouldn't the header provider be set on the top level settings? If I understand correctly, the header provider on the channel provider is set by the client context during build: 
https://github.com/googleapis/gax-java/blob/122eda65b4857cce676b53f07d19fef59dc0cb15/gax/src/main/java/com/google/api/gax/rpc/ClientContext.java#L151

Reply:
Unfortunately, while [creating `BigtableStubSettings` in `EnhancedBigtableStub#create`](https://github.com/rahulKQL/google-cloud-java/blob/master/google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java#L88-L94), we are not considering client’s `HeaderProvider` value present in the argument.

I cross checked,  only default header i.e. `x-goog-api-client` is available at ClientContext’s Xtor. We would need user-agent to be available at `TransportChannelProvider`, for that reason I created `InstantiatingGrpcChannelProvider`.
Also we had to set `ManagedChannelBuilder#usePlainText` to support plainText negotiation rpcs.

Do let me know If I should raise PR to consider HeaderProvider at `EnhancedBigtableStub#create` before ClientContext creation.

Reply:
Please do, its a bug: the header provider should be copied over in the enhanced stub

Reply:
[PR#4491](https://github.com/googleapis/google-cloud-java/pull/4491) for headerProvider addition is merged now. 

Once the next release(i.e. v0.81.0) is available, I will have this PR rebase and updated.